legacy/logclient/persistance.go

```diff
@@ -0,0 +1,130 @@
+*/
+
+package logclient
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// PersistentLoggingInfo holds peristand stats about the behavior of the auditor.
+// Since the auditor is a continuously running program, this data will begin
+// collection upon startup and lost on shutdown. Such data is only collected when
+// invoked with the --verbose flag.
+type PersistentLoggingInfo struct {
+	verboseLogging bool             // Determines if the user passed the --verbose flag, allowing persistent logs.
+	mutex          sync.Mutex       // Required to grab the queries after each measurement window.
+	measurement    QueryMeasurement // The measurement structure holding the amount of times GCR was queried.
+}
+
+// QueryMeasurement holds the logging information about number of queries per measurment window.
+type QueryMeasurement struct {
+	startTime  time.Time
+	recorded   bool
+	numQueries uint64
+}
+
+// Debug only logs the given message if the verboseLogging option is true.
+func Debug(args ...interface{}) {
+	if PersistentAuditInfo.verboseLogging {
+		logrus.Debug(args...)
+	}
+}
+
+// CountGCRQuery adds the given number to the persistent audit logs only
+// if verboseLogging is enabled.
+func CountGCRQuery(n int) {
+	// Only capture queries when verboseLogging is enabled
+	if !PersistentAuditInfo.verboseLogging {
+		return
+	}
+	// Grab and the lock to prevent race conditions.
+	PersistentAuditInfo.mutex.Lock()
+	defer PersistentAuditInfo.mutex.Unlock()
+	// Determin if the current measurement has already been recorded.
+	if !PersistentAuditInfo.measurement.recorded {
+		PersistentAuditInfo.measurement.numQueries += uint64(n)
+		return
+	}
+	// Create a new measurement.
+	PersistentAuditInfo.measurement = QueryMeasurement{
+		startTime:  time.Now(),
+		recorded:   false,
+		numQueries: uint64(n),
+	}
+}
+
+// queryLogger records the number of queries within the specified
+// measurement window.
+func queryLogger() {
+	// This logger must not terminate unless the program exists.
+	for {
+		// Wait for the query window.
+		time.Sleep(GCRMeasurementWindow)
+		// Grab the number of queries, and mark as recorded.
+		var numQueries uint64
+		PersistentAuditInfo.mutex.Lock()
+		numQueries = PersistentAuditInfo.measurement.numQueries
+		PersistentAuditInfo.measurement.recorded = true
+		PersistentAuditInfo.mutex.Unlock()
+		// Log the number of queries.
+		msg := fmt.Sprintf("Recorded %d Queries to GCR within %d min.", numQueries, GCRMeasurementWindow/time.Minute)
+		Debug(msg)
+		// Record when the auditor surpasses the GCR query limit.
+		if numQueries > GCRQueryLimit {
+			msg = fmt.Sprintf("Quota limit %d req every %d min surpassed. Number of queries recorded: %d", GCRQueryLimit, GCRMeasurementWindow/time.Minute, numQueries)
+			Debug(msg)
+		}
+	}
+}
+
+// EnablePersistentLogging configures the persistent logger to capture and
+// and display logging extra information.
+func EnablePersistentLogging() {
```

Comment:
I would recommend instead of having package level logging library methods; have different types of concrete logger structs. e.g. `PersistentLogger` which satisfy a common interface and create and use an instance of that (based on logic) and then pass that logger around.